lib/trace/index.js

```diff
@@ -0,0 +1,165 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var errors = require('../errors');
+var safeParse = require('../util').safeParse;
+var toCamelCase = require('../util').toCamelCase;
+
+var core = require('./core');
+var TChannelTracer = require('./tchannel');
+var LogTracer = require('./log');
+
+function registerTraceEndpoints(ringpop) {
+    // XXX testing only?
+    //
+    // @wolski, we chatted briefly about this -- whatcha think should
+    // be done? the testing framework doesn't work, as is. (I can add
+    // more mocks to make it go.)
+    if (!ringpop || !ringpop.channel || !ringpop.membership) {
+        return;
+    }
+
+    registerTraceCommand(ringpop, '/trace/add', createTracer);
+    registerTraceCommand(ringpop, '/trace/remove', destroyTracer);
+}
+
+function registerTraceCommand(ringpop, endpoint, processor) {
+    var logPrefix = 'ringpop' + endpoint.replace(/\//g, '.');
+
+    ringpop.channel.register(endpoint, function onRequest(req, resp, optsStr) {
+        resp.headers.as = 'raw';
+        var opts = toCamelCase(safeParse(optsStr));
+        var config = core.resolveTracerConfig(ringpop, opts.event);
+        if (!config) {
+            var err = errors.InvalidOptionError({ option: 'event', reason: 'it is invalid'});
+            ringpop.logger.warn(logPrefix + ': invalid event', err);
+            resp.sendNotOk(err);
+            return;
+        }
+
+        processor(ringpop, config, opts, function done(err) {
+            if (err) {
+                ringpop.logger.warn(logPrefix + ': remote error', err);
+                resp.sendNotOk(err);
+                return;
+            } 
+            ringpop.logger.info(logPrefix + ': success', opts);
+            resp.sendOk();
+        });
+    });
+}
+
+var knownTracerTypes = {
+    tchannel: true,
+    log: true
+};
+
+function createTracer(ringpop, config, opts, callback) {
+    callback = callback || function() {};
+
+    if (!core.Tracer.validate(opts) || !knownTracerTypes[opts.sink.type]) {
+        callback(errors.InvalidOptionError(
+            { option: 'sink.type', reason: 'is empty or unknown' }));
+        return;
+    }
+
+    if (indexOfTracer(ringpop, config, opts) >= 0) {
```

Comment:
Rather than the registered tracers hanging off of the ringpop instance, what are your thoughts about creating an explicit TracerStore aka `lib/trace/store.js` for registered tracers and thus you can call something like:

```
if (store.hasTracer(...)) {
  // idempotent
}
```